tools/clang/tools/dxil2spv/lib/dxil2spv.cpp

```diff
@@ -130,6 +124,153 @@ int Translator::Run(CComPtr<IDxcBlobEncoding> blob) {
+        &outputSignature) {
+  interfaceVars.reserve(inputSignature.size() + outputSignature.size());
+
+  // Translate DXIL input signature to SPIR-V stage input vars.
+  for (std::unique_ptr<hlsl::DxilSignatureElement> &elem : inputSignature) {
+    clang::spirv::SpirvVariable *var = spvBuilder.addStageIOVar(
+        spvContext.getPointerType(toSpirvType(elem.get()),
+                                  spv::StorageClass::Input),
+        spv::StorageClass::Input, elem->GetSemanticName(), false, {});
+    interfaceVars.push_back(var);
+    inputSignatureElementMap[elem->GetID()] = var;
+  }
+
+  // Translate DXIL input signature to SPIR-V stage ouput vars.
+  for (std::unique_ptr<hlsl::DxilSignatureElement> &elem : outputSignature) {
+    clang::spirv::SpirvVariable *var = spvBuilder.addStageIOVar(
+        spvContext.getPointerType(toSpirvType(elem.get()),
+                                  spv::StorageClass::Output),
+        spv::StorageClass::Output, elem->GetSemanticName(), false, {});
+    interfaceVars.push_back(var);
+    outputSignatureElementMap[elem->GetID()] = var;
+  }
+}
+
+void Translator::createEntryFunction(llvm::Function *entryFunction) {
+  spirv::SpirvFunction *spirvEntryFunction =
+      spvBuilder.beginFunction(toSpirvType(entryFunction->getReturnType()), {},
+                               entryFunction->getName());
+
+  // Translate entry function parameters.
+  llvm::SmallVector<const spirv::SpirvType *, 4> spirvParamTypes;
+  for (llvm::Argument &argument : entryFunction->getArgumentList()) {
+    spirvParamTypes.push_back(toSpirvType(argument.getType()));
+  }
+  spirvEntryFunction->setFunctionType(spvContext.getFunctionType(
+      spirvEntryFunction->getReturnType(), spirvParamTypes));
+
+  // Create entry function basic blocks.
+  for (llvm::BasicBlock &basicBlock : *entryFunction) {
+    createBasicBlock(basicBlock);
+  }
+  spvBuilder.endFunction();
+  spvBuilder.addEntryPoint(
+      spirv::SpirvUtils::getSpirvShaderStage(
+          spvContext.getCurrentShaderModelKind()),
+      spirvEntryFunction, spirvEntryFunction->getFunctionName(), interfaceVars);
+}
+
+void Translator::createBasicBlock(llvm::BasicBlock &basicBlock) {
+  spvBuilder.setInsertPoint(spvBuilder.createBasicBlock());
+
+  // Translate each intruction in basic block.
+  for (llvm::Instruction &instruction : basicBlock) {
+    createInstruction(instruction);
+  }
+}
+
+void Translator::createInstruction(llvm::Instruction &instruction) {
+  // Handle Call instructions.
+  if (isa<llvm::CallInst>(instruction)) {
+    llvm::CallInst &callInstruction = cast<llvm::CallInst>(instruction);
+    // Many shader operations are represented by call instructions that call
+    // external functions, whose name is prefixed with dx.op. (i.e.
+    // dx.op.loadInput, dx.op.storeOutput).
+    hlsl::DXIL::OpCode dxilOpcode =
+        hlsl::OP::GetDxilOpFuncCallInst(&instruction);
+    switch (dxilOpcode) {
+    case hlsl::DXIL::OpCode::LoadInput: {
```

Comment:
nit:
This method will become very long and complicated in long term. I'd like to split each case as a method e.g., `createLoadInstructionForInput()`.

Reply:
Done